### PR TITLE
AddOp: Remove HLO_CompatibleOperandsAndResultType Trait, add custom verifier

### PR DESCRIPTION
**Pull request by abhigunj**
AddOP allows mix of `per-tensor` and `per-axis` inputs. Existing StableHLO AddOp def uses `HLO_CompatibleOperandsAndResultType` Trait which does not allow mix of `per-tensor` and `per-axis` inputs. 
This PR:
removed HLO_CompatibleOperandsAndResultType Trait for AddOP
Added custom Verifier to implement OP constraints

Thank you @sdasgup3  for the help with debugging builder issues after removal of `HLO_CompatibleOperandsAndResultType` suggestion for  `inferReturnTypes` placement.

### CODE REVIEW

File: stablehlo/tests/ops_stablehlo.mlir [hunk 1]
```diff
@@ -5058,62 +5058,85 @@ func.func @is_compatible_dynamism_mix(%arg0: tensor<?xf32>, %arg1: tensor<1xf32>
...
+  // expected-error@+1 {{quantization_dimension of rhs and result are not same}}
+  %0 = "stablehlo.add"(%arg0, %arg1) : (tensor<1x2x!quant.uniform<i8:f32:0, {1.0:17}>>, tensor<1x2x!quant.uniform<i8:f32:1, {1.0:17}>>) -> tensor<1x2x!quant.uniform<i8:f32:0, {1.0:17}>>
   func.return
 }
 
 // -----
 
 func.func @is_compatible_quant_signedness_mismatch(%arg0: tensor<1x!quant.uniform<i8:f32, 1.0:17>>) {
-  // expected-error@+1 {{op requires compatible types for all operands and results}}
+  // expected-error@+2 {{'stablehlo.add' op failed to infer returned types}}
```
Comment (sdasgup3) [inline stablehlo/tests/ops_stablehlo.mlir:5138]: remove `stablehlo.add` from error check

File: stablehlo/tests/stablehlo_refine_shapes.mlir [hunk 2]
```diff
@@ -3,7 +3,7 @@
 func.func @error_illformed(%arg0: tensor<3xf32>, %arg1: tensor<4xf32>) -> tensor<?xf32> {
   %0 = stablehlo.abs %arg0 : (tensor<3xf32>) -> tensor<?xf32>
   %1 = stablehlo.abs %arg1 : (tensor<4xf32>) -> tensor<?xf32>
-  // expected-error@+1{{requires compatible types for all operands and results}}
+  // expected-error@+1{{requires the same shape for all operands and results}}
```
Comment (sdasgup3) [inline stablehlo/tests/stablehlo_refine_shapes.mlir:6]: [no action needed]

liked that the error messages are more informative.

File: stablehlo/dialect/TypeInference.cpp [hunk 3]
```diff
@@ -264,6 +281,76 @@ LogicalResult verifyPairwiseCompatibleShapes(TypeRange values) {
...
 }
 
+LogicalResult verifyAddOp(std::optional<Location> location, Type lhsType,
+                          Type rhsType, Type resultType) {
+  lhsType = getElementTypeOrSelf(lhsType);
+  rhsType = getElementTypeOrSelf(rhsType);
+  resultType = getElementTypeOrSelf(resultType);
+  llvm::SmallVector<Type, 3> typeEntries{lhsType, rhsType, resultType};
+  // add_c1
+  if (noneQuantized<quant::QuantizedType>(typeEntries)) {
```
Comment (GleasonK) [inline stablehlo/dialect/TypeInference.cpp:291]: My concern with this change is -- not all users of StableHLO want to think about quantization. It's a subset of total users who use this feature, but they'd have to filter through a very complicated implementation of verifyAdd to understand what their uses need. In that sense, perhaps we can write this function as:

```
if (anyQuantized(lhsType,rhsType,resultType))
  return verifyBinaryOpQuantizationConstraints(...);

return OpTrait::SameOperandsAndResultType<AddOp>::verifyTrait(getOperation());
```

**Review by GleasonK — CHANGES_REQUESTED** [review]

File: stablehlo/dialect/TypeInference.cpp (hunk 3 above)
Comment (abhigunj) [reply stablehlo/dialect/TypeInference.cpp:291]: Yeh.. agree. I will put the Q Verifier checks under dedicated verifyBinaryOpQuantizationConstraints function.

File: stablehlo/dialect/TypeInference.cpp [hunk 4]
```diff
@@ -65,6 +65,100 @@ limitations under the License.
...
+
+LogicalResult verifyBinaryOpQuantizationConstraints(
+    std::optional<Location> location, Type lhsType, Type rhsType,
+    Type resultType) {
+  lhsType = getElementTypeOrSelf(lhsType);
+  rhsType = getElementTypeOrSelf(rhsType);
+  resultType = getElementTypeOrSelf(resultType);
+  llvm::SmallVector<Type, 3> typeEntries{lhsType, rhsType, resultType};
+  // add_c1
+  if (noneQuantized<quant::QuantizedType>(typeEntries)) {
```
Comment (sdasgup3) [inline stablehlo/dialect/TypeInference.cpp:98]: This has already been checked at `verifyAddOp`? So you can remove this if

**Review by sdasgup3 — APPROVED** [review]

**Review by GleasonK — APPROVED** [review]

File: stablehlo/dialect/TypeInference.cpp (hunk 3 above)
Comment (abhigunj) [reply stablehlo/dialect/TypeInference.cpp]: for : `return OpTrait::SameOperandsAndResultType<AddOp>::verifyTrait(getOperation());`

AddOp is  `StableHLO_BinaryElementwiseOp` and shape check is enabled by default. Custom verifier will check  same element type only.
tried  using
`mlir::OpTrait::SameOperandsAndResultElementType<::mlir::stablehlo::AddOp>::verifyTrait(op);` but bazel build failed. There is a circular dependency if I try to add`stablehlo_ops` to `stablehlo_type_inference` target to use `::mlir::stablehlo::AddOp`  at TypeInference.cpp. 

instead trait will do element type check using
`lhsType == resultType ` `lhsType == rhsType`

Comment (GleasonK) [reply stablehlo/dialect/TypeInference.cpp]: Can we use `SameOperandsAndResultElementType <Operation*>`? Otherwise we can use equality and our own mesage

Comment (abhigunj) [reply stablehlo/dialect/TypeInference.cpp]: Thank you. yeh.. tried that too before using quality and our own message.

File: stablehlo/dialect/TypeInference.cpp [hunk 5]
```diff
@@ -65,6 +65,94 @@ limitations under the License.
...
+  auto rhsQPAType = rhsType.dyn_cast<quant::UniformQuantizedPerAxisType>();
+  auto resultQPAType =
+      resultType.dyn_cast<quant::UniformQuantizedPerAxisType>();
+  if (lhsQPAType || rhsQPAType) {
+    // add_c5
+    if (!resultQPAType)
+      return emitOptionalError(
+          location, "result is not per_axis quantized but lhs or rhs are");
+    // add_c6
+    if (lhsQPAType)
```
Comment (GleasonK) [inline stablehlo/dialect/TypeInference.cpp:131]: Add curly braces to outer if to avoid dangling else

File: stablehlo/dialect/TypeInference.cpp [hunk 6]
```diff
@@ -65,6 +65,94 @@ limitations under the License.
...
+      if (resultQPAType.getQuantizedDimension() !=
+          lhsQPAType.getQuantizedDimension())
+        return emitOptionalError(
+            location, "quantization_dimension of lhs and result are not same ",
+            lhsType, " vs ", resultType);
+    // add_c7
+    if (rhsQPAType)
```
Comment (GleasonK) [inline stablehlo/dialect/TypeInference.cpp:138]: Add curly braces to outer if to avoid dangling else

File: stablehlo/dialect/TypeInference.cpp [hunk 7]
```diff
@@ -65,6 +65,94 @@ limitations under the License.
...
+      if (resultQPAType.getQuantizedDimension() !=
+          rhsQPAType.getQuantizedDimension())
+        return emitOptionalError(
+            location, "quantization_dimension of rhs and result are not same ",
+            rhsType, " vs ", resultType);
+
+    return success();
+  }
+
+  return !resultQPAType
```
Comment (GleasonK) [inline stablehlo/dialect/TypeInference.cpp:148]: Let's just use:

```
if (resultQPAType)
  emitOptionaleError(...);

return success();
```

Easier readability

**Review by GleasonK — APPROVED** [review]

**Pull request merged**